docs/original_docs/configuration.md
# Configuration and Parameters

FlashRAG enables comprehensive management of various parameters for controlling the experiment. FlashRAG supports two types of parameter configurations: **YAML config file** and **parameter dict**. The parameters are assigned via the `Config` module.

In practice, `config` can be used in a dictionary access manner, such as:
```python
from flashrag.config import Config
config_dict = {'generator_model': 'llama2-7B'}
config = Config(config_dict=config_dict)
model_name = config['generator_model']
```

In most cases, the majority of parameters do not need to be modified. Specific application examples can be found below in the [Example Configuration](#example-configuration).

## Operation Logic

##### Config File

Config File should be in YAML format, and users should fill in the corresponding parameters according to the YAML syntax. In our library, we provide a template file for users to refer to, with comments explaining the specific meaning of each parameter.

The code for using a config file is as follows:
```python 
from flashrag.config import Config
config = Config(config_file_path='myconfig.yaml')
```

##### Parameter Dict

Another way is to set up the configuration via a Python dictionary, where keys are parameter names and values are parameter values. This method is more flexible compared to using a file.

The code for using a parameter dict is as follows:
```python 
from flashrag.config import Config
config_dict = {'generator_model': 'llama2-7B'}
config = Config(config_dict=config_dict)
```

##### Priority

In FlashRAG, we support combining both methods.

The priority of configuration methods is: Parameter Dict > Config File > Default Settings

The default settings are recorded in [basic_config.yaml](../flashrag/config/basic_config.yaml).

##### Path Settings

To facilitate the management of various model paths and index paths, we introduce `model2path`, `model2pooling`, and `method2index`. Users only need to fill in the corresponding names and paths here to automatically load the paths to be used.

For example, after setting the path of `llama2-7B` in `model2path`, you can directly specify it as the corresponding model without rewriting the path.
```yaml
model2path:
    llama2-7B: 'my_path'
generator_model: 'llama2-7B'
```

If the corresponding model path is not found in the dictionary, it will look for the parameter corresponding to the model path (e.g., `generator_model_path`).

This rule applies to the following parameters:
- retrieval_method
- rerank_model_name
- generator_model
- index_path

## Example Configuration

An example configuration file can be found in [basic_config.yaml](../flashrag/config/basic_config.yaml). Below we will go through each section in the configuration file.

```yaml
# basic settings

# ------------------------------------------------Global Paths------------------------------------------------#
# Paths to various models
model2path:
  e5: "intfloat/e5-base-v2"
  bge: "intfloat/e5-base-v2"
  contriever: "facebook/contriever"
  llama2-7B-chat: "meta-llama/Llama-2-7b-chat-hf"
  llama2-7B: "meta-llama/Llama-2-7b-hf"
  llama2-13B: "meta-llama/Llama-2-13b-hf"
  llama2-13B-chat: "meta-llama/Llama-2-13b-chat-hf"
  
# Pooling methods for each embedding model
model2pooling:
  e5: "mean"
  bge: "cls"
  contriever: "mean"
  jina: 'mean'
  dpr: cls

# Indexes path for retrieval models
method2index:
  e5: ~
  bm25: ~
  contriever: ~

# ------------------------------------------------Environment Settings------------------------------------------------#
# Directory paths for data and outputs
data_dir: "datasets/"
save_dir: "output/"

gpu_id: "0,1,2,3"
dataset_name: "nq" # name of the datasets in data_dir
split: ["test"]  # datasets split to load (e.g. train,dev,test)

# Sampling configurations for testing
test_sample_num: ~  # number of samples to test (only work in dev/test split), if None, test all samples
random_sample: False # whether to randomly sample the test samples

# Seed for reproducibility
seed: 2024

# Whether save intermediate data
save_intermediate_data: True
save_note: 'experiment'

# -------------------------------------------------Retrieval Settings------------------------------------------------#
# If set the name, the model path will be find in global paths
retrieval_method: "e5"  # name or path of the retrieval model. 
retrieval_model_path: ~ # path to the retrieval model
index_path: ~ # set automatically if not provided. 
faiss_gpu: False # whether use gpu to hold index
corpus_path: ~  # path to corpus in '.jsonl' format that store the documents

use_sentence_transformer: False # If set, the retriever will be load through `sentence transformer` library
retrieval_topk: 5 # number of retrieved documents
retrieval_batch_size: 256  # batch size for retrieval
retrieval_use_fp16: True  # whether to use fp16 for retrieval model
retrieval_query_max_length: 128  # max length of the query
save_retrieval_cache: True # whether to save the retrieval cache
use_retrieval_cache: False # whether to use the retrieval cache
retrieval_cache_path: ~ # path to the retrieval cache
retrieval_pooling_method: ~ # set automatically if not provided

use_reranker: False # whether to use reranker
rerank_model_name: ~ # same as retrieval_method
rerank_model_path: ~ # path to reranker model, path will be automatically find in `model2path`
rerank_pooling_method: ~
rerank_topk: 5  # number of remain documents after reranking
rerank_max_length: 512 
rerank_batch_size: 256 # batch size for reranker
rerank_use_fp16: True

# -------------------------------------------------Generator Settings------------------------------------------------#
framework: fschat # inference frame work of LLM, supporting: 'hf','vllm','fschat', 'openai'
generator_model: "llama3-8B-instruct" # name or path of the generator model
# setting for openai model, only valid in openai framework
openai_setting:
  api_key: ~
  base_url: ~

generator_model_path: ~
generator_max_input_len: 1024  # max length of the input
generator_batch_size: 4 # batch size for generation, invalid for vllm
generation_params:  
  #do_sample: false
  max_tokens: 32
  #temperature: 1.0
  #top_p: 1.0
use_fid: False # whether to use FID, only valid in encoder-decoder model

# -------------------------------------------------Evaluation Settings------------------------------------------------#
# Metrics to evaluate the result
metrics: ['em','f1','acc,'precision','recall'] 
# Specify setting for metric, will be called within certain metrics
metric_setting: 
  retrieval_recall_topk: 5
save_metric_score: True #　whether to save the metric score into txt file
```


### Global Settings

Here the paths to models, indexes, and pooling methods for each embedding model are saved. Later, you only need to specify the corresponding name to automatically load the path.

```yaml
# Paths to various models
model2path:
  e5: "intfloat/e5-base-v2"
  bge: "intfloat/e5-base-v2"
  contriever: "facebook/contriever"
  llama2-7B-chat: "meta-llama/Llama-2-7b-chat-hf"
  llama2-7B: "meta-llama/Llama-2-7b-hf"
  llama2-13B: "meta-llama/Llama-2-13b-hf"
  llama2-13B-chat: "meta-llama/Llama-2-13b-chat-hf"
  
# Pooling methods for each embedding model
model2pooling:
  e5: "mean"
  bge: "cls"
  contriever: "mean"
  jina: 'mean'
  dpr: cls

# Indexes path for retrieval models
method2index:
  e5: ~
  bm25: ~
  contriever: ~
```

### Environment Settings

Here mainly manage various configurations of the experiment.

```yaml
# Directory paths for data and outputs
data_dir: "datasets/"
save_dir: "output/"

gpu_id: "0,1,2,3"
dataset_name: "nq" # name of the datasets in data_dir
split: ["test"]  # datasets split to load (e.g. train,dev,test)

# Sampling configurations for testing
test_sample_num: ~  # number of samples to test (only work in dev/test split), if None, test all samples
random_sample: False # whether to randomly sample the test samples

# Seed for reproducibility
seed: 2024

# Whether save intermediate data
save_intermediate_data: True
save_note: 'experiment'
```

- `split`: Specifies the dataset split to load, multiple splits can be specified and used separately in the code.
- `save_note`, `save_dir`: Each experiment will create a new folder in `save_dir`, and add `save_note` as a marker in the folder name.
- `save_intermediate_data`: If enabled, intermediate results will be recorded in the experiment folder, including the retrieval content, generated content, and corresponding evaluation metrics for the content.
- `gpu_id`: Specifies which GPUs to use for the experiment, supports multiple GPUs.

If you do not want to use the entire dataset for testing, you can set 'test_sample_num' (invalid for the training set). If set, the corresponding number of samples at the front will be selected. If 'random_sample' is enabled, it will be randomly selected.

### Retrieval Settings

This section manages various parameters for the retriever and reranker.

```yaml
retrieval_method: "e5"  # name or path of the retrieval model. 
retrieval_model_path: ~ # path to the retrieval model
index_path: ~ # set automatically if not provided. 
faiss_gpu: False # whether use gpu to hold index
corpus_path: ~  # path to corpus in '.jsonl' format that store the documents

use_sentence_transformer: False # If set, the retriever will be load through `sentence transformer` library
retrieval_topk: 5 # number of retrieved documents
retrieval_batch_size: 256  # batch size for retrieval
retrieval_use_fp16: True  # whether to use fp16 for retrieval model
retrieval_query_max_length: 128  # max length of the query
save_retrieval_cache: True # whether to save the retrieval cache
use_retrieval_cache: False # whether to use the retrieval cache
retrieval_cache_path: ~ # path to the retrieval cache
retrieval_pooling_method: ~ # set automatically if not provided

use_reranker: False # whether to use reranker
rerank_model_name: ~ # same as retrieval_method
rerank_model_path: ~ # path to reranker model, path will be automatically find in `model2path`
rerank_pooling_method: ~
rerank_topk: 5  # number of remain documents after reranking
rerank_max_length: 512 
rerank_batch_size: 256 # batch size for reranker
rerank_use_fp16: True
```


If the paths in the previous dictionary are filled, only `retrieval_method` and `corpus_path` need to be modified (no need to modify `index_path` and `retrieval_model_path`).

FlashRAG supports saving and reusing retrieval results. When reusing, it will look in the cache to see if there is a query identical to the current one and read the corresponding results.
- `save_retrieval_cache`: If set to `True`, it will save the retrieval results as a JSON file, recording the retrieval results and scores for each query, enabling reuse next time.
- `retrieval_cache_path`: Set to the path of the previously saved retrieval cache.

To use a reranker, set `use_reranker` to `True` and fill in `rerank_model_name`. For Bi-Embedding type rerankers, the pooling method needs to be set, similar to the retrieval method.

If set `use_sentence_transformer` to `True`, there is no need to set consider pooling method.

### Generator Settings

This section records various settings for the generator.

```yaml
framework: fschat # inference frame work of LLM, supporting: 'hf','vllm','fschat', 'openai'
generator_model: "llama3-8B-instruct" # name or path of the generator model
# setting for openai model, only valid in openai framework
openai_setting:
  api_key: ~
  base_url: ~
generator_model_path: ~
generator_max_input_len: 1024  # max length of the input
generator_batch_size: 4 # batch size for generation, invalid for vllm
generation_params:  
  max_tokens: 32
use_fid: False # whether to use FID, only valid in encoder-decoder model
```

- `framework`: The base framework of the generator. It is recommended to use `vllm` for deployment.
- `generation_params`: Parameters needed during generation. The parameter names may need to be adjusted according to different frameworks. Refer to the function descriptions of vllm or huggingface generation for details.

### Evaluation Settings

This section sets various settings used during evaluation. If you use a custom evaluation metric, you can add your parameters in `metric_setting` and call them in the metric.

```yaml
# Metrics to evaluate the result
metrics: ['em','f1','acc','precision','recall'] 
# Specify setting for metric, will be called within certain metrics
metric_setting: 
  retrieval_recall_topk: 5
save_metric_score: True #　whether to save the metric score into txt file
```

- `metrics`: The specific evaluation metrics to be used. The values are the `metric_name` of the evaluation metrics. Currently supported evaluation metrics can be found [<u>here</u>](../flashrag/evaluator/metrics.py).

docs/zh-cn/configuration/basic-settings.md
# 环境设置

环境设置主要包含与组件无关的一些全局设置，包括自动路径设置，全局设置和评测设置。

## 自动路径设置

为了方便管理各种模型路径和索引路径，我们引入了 `model2path`、`model2pooling` 和 `method2index`。用户只需在这里填写相应的名称和路径，即可自动加载所需路径。

例如，在 model2path 中设置了 llama2-7B 的路径后，可以直接在`generator_model`里面填写`llama2-7B`,无需再填写`generator_model_path`，可以简化各类路径的管理。
```yaml
model2path:
    llama2-7B: 'my_path'
generator_model: 'llama2-7B'
```

如果在`model2path`中没有找到`generator_model`对应的键，则需要填写`generator_model_path`作为模型路径。

其识别关系如下:

* **model2path**: 识别`generator_model`，`retrieval_method`和`rerank_model_name`，并自动填写对应模型路径。
* **model2pooling**: 识别`retrieval_method`，并填入`retrieval_pooling_method`。(当前版本pooling已经能够自动填写，不需要设置)
* **index2path**: 识别`retrieval_method`，并填入`index_path`。

## 全局设置

全局设置包括了实验环境以及数据集加载的各种配置。涉及全局设置的参数如下:

```yaml
# 数据集加载路径以及实验保存路径
data_dir: "datasets/"
save_dir: "output/"

gpu_id: "0,1,2,3"
dataset_name: "nq" # 数据集名称
split: ["test"] # 需要加载的数据集子集

test_sample_num: ~ # 要测试的样本数（仅在 dev/test 分割中有效），如果为 None，则测试所有样本
random_sample: False # 是否随机采样测试样本

# 随机种子
seed: 2024

# 是否保存中间数据
save_intermediate_data: True
save_note: "experiment"
```

其中重要参数的含义如下：
* `data_dir`:  存放所有数据集的路径，加载的数据集需存放在`data_dir/dataset_name路径下
* `split`：指定要加载的数据集子集(train/dev/test), 所有被指定的子集都会被加载。
* `save_note`, `save_dir`：每次初始化`Config`都将在 `save_dir` 中创建一个新文件夹，并在文件夹名称中添加 `save_note` 作为标记。文件夹会保存实验涉及的参数、中间数据和评测结果。
* `save_intermediate_data`：如果启用，中间结果将记录在实验文件夹中，包括检索内容、生成内容等。
* gpu_id：指定实验使用哪些 GPU，支持多 GPU。


## 评测设置

评估设置部分定义了实验中用来衡量模型表现的各种评估指标以及其他相关配置，帮助在实验过程中进行结果评估。以下是该部分配置文件的详细说明：

```yaml
# 用到的评测指标
metrics: ["em", "f1", "acc", "precision", "recall", "input_tokens"]
# 评测指标中涉及的特殊设置
metric_setting:
  retrieval_recall_topk: 5
  tokenizer_name: "gpt-4"
# 是否保存最终的评测分数
save_metric_score: True 
```

参数含义如下:

* metrics: 评估方法时使用的评测指标指标列表。目前支持的评测指标包括：
     - `em`: exact match，即模型输出结果与标准答案是否完全一致。
     - `acc`: accuracy，即模型输出结果是否包含正确答案。
     - `f1`: F1 分数，衡量输出答案和标准答案的token-level分数。
     - `precision`: Precision 分数，衡量输出答案和标准答案的token-level 分数。
     - `recall`: recall 分数，衡量输出答案和标准答案的token-level分数。
     - `input_tokens`: 输入模型的 tokens 的数量，用于记录每个样本的输入长度。
     - `retrieval_recall`: 检索的召回率，衡量检索文档是否包含正确答案。
     - `retrieval_precision`: 检索的精确性
     - `rouge-1`, `rouge-2`, `rouge-l`
     - `zh_rouge-1`, `zh_rouge-2`, `zh_rouge-l`: 中文版本
     - `bleu`
     - `gaokao_acc`: 评估GAOKAO-MM数据集的各科目Accuracy
     - `llm_judge`: 基于LLM对生成答案进行打分，需要额外配置`llm_judge_setting`。

* metric_setting: 该部分定义了与评估指标相关的额外设置：
     - `retrieval_recall_topk`: 在检索任务中，指定需要考虑的召回结果数量，这里设置为 5，意味着评估时考虑前 5 个候选结果。
     - `tokenizer_name`: 用于计算输入token数量，可指定openai系列的分词器以及HF支持的各类tokenizer。

* save_metric_score: 是否将评估结果保存到txt文件中。设置为 `True` 表示会将每个评估指标的得分保存在实验文件夹下的`metric_score.txt`文件中。



flashrag/config/config.py
import re
import os
os.environ["TOKENIZERS_PARALLELISM"] = "false"
import yaml
import random
import datetime


class Config:
    def __init__(self, config_file_path=None, config_dict={}):

        self.yaml_loader = self._build_yaml_loader()
        self.file_config = self._load_file_config(config_file_path)
        self.variable_config = config_dict

        self.external_config = self._merge_external_config()

        self.internal_config = self._get_internal_config()

        self.final_config = self._get_final_config()

        self._check_final_config()
        self._set_additional_key()

        self._init_device()
        self._set_seed()
        if not self.final_config.get('disable_save', False):
            self._prepare_dir()

    def _build_yaml_loader(self):
        loader = yaml.FullLoader
        loader.add_implicit_resolver(
            "tag:yaml.org,2002:float",
            re.compile(
                """^(?:
             [-+]?(?:[0-9][0-9_]*)\\.[0-9_]*(?:[eE][-+]?[0-9]+)?
            |[-+]?(?:[0-9][0-9_]*)(?:[eE][-+]?[0-9]+)
            |\\.[0-9_]+(?:[eE][-+][0-9]+)?
            |[-+]?[0-9][0-9_]*(?::[0-5]?[0-9])+\\.[0-9_]*
            |[-+]?\\.(?:inf|Inf|INF)
            |\\.(?:nan|NaN|NAN))$""",
                re.X,
            ),
            list("-+0123456789."),
        )
        return loader

    def _load_file_config(self, config_file_path: str):
        file_config = dict()
        if config_file_path:
            with open(config_file_path, "r", encoding="utf-8") as f:
                file_config.update(yaml.load(f.read(), Loader=self.yaml_loader))
        return file_config

    @staticmethod
    def _update_dict(old_dict: dict, new_dict: dict):
        # Update the original update method of the dictionary:
        # If there is the same key in `old_dict` and `new_dict`, and value is of type dict, update the key in dict

        same_keys = []
        for key, value in new_dict.items():
            if key in old_dict and isinstance(value, dict):
                same_keys.append(key)
        for key in same_keys:
            old_item = old_dict[key]
            new_item = new_dict[key]
            old_item.update(new_item)
            new_dict[key] = old_item

        old_dict.update(new_dict)
        return old_dict

    def _merge_external_config(self):
        external_config = dict()
        external_config = self._update_dict(external_config, self.file_config)
        external_config = self._update_dict(external_config, self.variable_config)

        return external_config

    def _get_internal_config(self):
        current_path = os.path.dirname(os.path.realpath(__file__))
        init_config_path = os.path.join(current_path, "basic_config.yaml")
        internal_config = self._load_file_config(init_config_path)

        return internal_config

    def _get_final_config(self):
        final_config = dict()
        final_config = self._update_dict(final_config, self.internal_config)
        final_config = self._update_dict(final_config, self.external_config)

        return final_config

    def _check_final_config(self):
        # check split
        split = self.final_config["split"]
        if split is None:
            split = ["train", "dev", "test"]
        if isinstance(split, str):
            split = [split]
        self.final_config["split"] = split

    def _init_device(self):
        gpu_id = self.final_config["gpu_id"]
        if gpu_id is not None:
            os.environ["CUDA_VISIBLE_DEVICES"] = str(gpu_id)
        try:
            # import pynvml 
            # pynvml.nvmlInit()
            # gpu_num = pynvml.nvmlDeviceGetCount()
            import torch
            gpu_num = torch.cuda.device_count()
        except:
            gpu_num = 0
        self.final_config['gpu_num'] = gpu_num
        if gpu_num > 0:
            self.final_config["device"] = "cuda"
        else:
            self.final_config['device'] = 'cpu'

    def _set_additional_key(self):
        def set_pooling_method(method, model2pooling):
            for key, value in model2pooling.items():
                if key.lower() in method.lower():
                    return value
            return "mean"

        def set_retrieval_keys(model2path, model2pooling, method2index, config):
            retrieval_method = config["retrieval_method"]
            if config["index_path"] is None:
                try:
                    config["index_path"] = method2index[retrieval_method]
                except:
                    print("Index is empty!!")

            if config.get("retrieval_model_path") is None:
                config["retrieval_model_path"] = model2path.get(retrieval_method, retrieval_method)

            if config.get("retrieval_pooling_method") is None:
                config["retrieval_pooling_method"] = set_pooling_method(retrieval_method, model2pooling)

            rerank_model_name = config.get("rerank_model_name", None)
            if config.get("rerank_model_path", None) is None:
                if rerank_model_name is not None:
                    config["rerank_model_path"] = model2path.get(rerank_model_name, rerank_model_name)
            if config.get("rerank_pooling_method", None) is None:
                if rerank_model_name is not None:
                    config["rerank_pooling_method"] = set_pooling_method(rerank_model_name, model2pooling)
            return config

        # set datasets
        dataset_name = self.final_config["dataset_name"]
        data_dir = self.final_config["data_dir"]
        self.final_config["dataset_path"] = os.path.join(data_dir, dataset_name)

        # set retrieval-related keys
        model2path = self.final_config["model2path"]
        model2pooling = self.final_config["model2pooling"]
        method2index = self.final_config["method2index"]
        self.final_config = set_retrieval_keys(model2path, model2pooling, method2index, self.final_config)
        # set keys for multi retriever
        if "multi_retriever_setting" in self.final_config:
            multi_retriever_config = self.final_config["multi_retriever_setting"]
            retriever_config_list = multi_retriever_config.get("retriever_list", [])
            # set for reranker merge method
            assert multi_retriever_config['merge_method'] in ['concat', 'rrf', 'rerank', None]
            if multi_retriever_config['merge_method'] == 'rerank':
                rerank_model_name = multi_retriever_config.get("rerank_model_name", None)
                assert rerank_model_name is not None
                multi_retriever_config['rerank_max_length'] = multi_retriever_config.get("rerank_max_length", 512)
                multi_retriever_config['rerank_batch_size'] = multi_retriever_config.get("rerank_batch_size", 256)
                multi_retriever_config['rerank_use_fp16'] = multi_retriever_config.get("rerank_use_fp16", True)
                
                if multi_retriever_config.get("rerank_model_path", None) is None:
                    if rerank_model_name is not None:
                        multi_retriever_config["rerank_model_path"] = model2path.get(rerank_model_name, rerank_model_name)
                if multi_retriever_config.get("rerank_pooling_method", None) is None:
                    if rerank_model_name is not None:
                        multi_retriever_config["rerank_pooling_method"] = set_pooling_method(rerank_model_name, model2pooling)
            
            # set config for each retriever
            for retriever_config in retriever_config_list:
                if "instruction" not in retriever_config:
                    retriever_config["instruction"] = None
                if "bm25_backend" not in retriever_config:
                    retriever_config["bm25_backend"] = "bm25s"
                if "use_reranker" not in retriever_config:
                    retriever_config["use_reranker"] = False
                if "index_path" not in retriever_config:
                    retriever_config["index_path"] = None
                if "corpus_path" not in retriever_config:
                    retriever_config["corpus_path"] = None
                if "use_sentence_transformer" not in retriever_config:
                    retriever_config["use_sentence_transformer"] = False
                retriever_config = set_retrieval_keys(model2path, model2pooling, method2index, retriever_config)
                
                # set other necessary keys as base setting
                keys = [
                    "retrieval_use_fp16",
                    "retrieval_query_max_length",
                    "faiss_gpu",
                    "retrieval_topk",
                    "retrieval_batch_size",
                    "use_reranker",
                    "rerank_model_name",
                    "rerank_model_path",
                    "retrieval_cache_path",
                ]
                for key in keys:
                    if key not in retriever_config:
                        retriever_config[key] = self.final_config.get(key, None)
                retriever_config["save_retrieval_cache"] = False
                retriever_config["use_retrieval_cache"] = False
        
        # set model path
        generator_model = self.final_config["generator_model"]

        if self.final_config.get("generator_model_path") is None:
            self.final_config["generator_model_path"] = model2path.get(generator_model, generator_model)

        if "refiner_name" in self.final_config:
            refiner_model = self.final_config["refiner_name"]
            if "refiner_model_path" not in self.final_config or self.final_config["refiner_model_path"] is None:
                self.final_config["refiner_model_path"] = model2path.get(refiner_model, None)
        if "instruction" not in self.final_config:
            self.final_config["instruction"] = None

        # set model path in metric setting
        metric_setting = self.final_config["metric_setting"]
        metric_tokenizer_name = metric_setting.get("tokenizer_name", None)
        from flashrag.utils.constants import OPENAI_MODEL_DICT

        if metric_tokenizer_name not in OPENAI_MODEL_DICT:
            metric_tokenizer_name = model2path.get(metric_tokenizer_name, metric_tokenizer_name)
            metric_setting["tokenizer_name"] = metric_tokenizer_name
            self.final_config["metric_setting"] = metric_setting

    def _prepare_dir(self):
        save_note = self.final_config["save_note"]
        save_dir = self.final_config['save_dir']
        if not save_dir.endswith("/"):
            save_dir += "/"

        current_time = datetime.datetime.now()

        self.final_config["save_dir"] = os.path.join(
            save_dir,
            f"{self.final_config['dataset_name']}_{current_time.strftime('%Y_%m_%d_%H_%M')}_{save_note}",
        )
        os.makedirs(self.final_config["save_dir"], exist_ok=True)
        # save config parameters
        config_save_path = os.path.join(self.final_config["save_dir"], "config.yaml")
        with open(config_save_path, "w") as f:
            yaml.dump(self.final_config, f, indent=4, sort_keys=False)

    def _set_seed(self):
        import torch
        import numpy as np
        seed = self.final_config['seed']
        try:
            seed = int(seed)
        except:
            seed = 2025
        self.final_config['seed'] = seed
        random.seed(seed)
        np.random.seed(seed)
        torch.manual_seed(seed)
        torch.cuda.manual_seed(seed)
        torch.cuda.manual_seed_all(seed)
        torch.backends.cudnn.benchmark = False
        torch.backends.cudnn.deterministic = True

    def __setitem__(self, key, value):
        if not isinstance(key, str):
            raise TypeError("index must be a str.")
        self.final_config[key] = value

    def __getattr__(self, item):
        if "final_config" not in self.__dict__:
            raise AttributeError("'Config' object has no attribute 'final_config'")
        if item in self.final_config:
            return self.final_config[item]
        raise AttributeError(f"'Config' object has no attribute '{item}'")

    def __getitem__(self, item):
        return self.final_config.get(item)

    def __contains__(self, key):
        if not isinstance(key, str):
            raise TypeError("index must be a str.")
        return key in self.final_config

    def __repr__(self):
        return self.final_config.__str__()
